ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs

```diff
@@ -347,11 +350,32 @@ forkBlockProduction IS{..} =
       case leaderResult of
         NotLeader ->
           return ()
-        ProducedBlock newBlock -> do
-          trace $ TraceForgeEvent currentSlot newBlock
-          ChainDB.addBlock chainDB newBlock
         FailedToProduce err ->
           trace $ TraceCouldNotForge currentSlot err
+        ProducedBlock newBlock txs -> do
+          trace $ TraceForgeEvent currentSlot newBlock
+          -- Adding a block is synchronous
+          ChainDB.addBlock chainDB newBlock
+          -- Check whether we adopted our block
+          curTip <- atomically $ ChainDB.getTipPoint chainDB
+          if curTip == blockPoint newBlock then
+            trace $ TraceAdoptedBlock currentSlot newBlock
+          else do
+            trace $ TraceDidntAdoptBlock currentSlot newBlock
+            isInvalid <- atomically $
+              ($ blockHash newBlock) . forgetFingerprint <$>
+              ChainDB.getIsInvalidBlock chainDB
+            whenJust isInvalid $ \reason -> do
```

Comment:
Do this check first, if it is invalid, trace `TraceForgedInvalidBlock`, it is is valid, trace `TraceDidntAdoptBlock` (include in the documentation of this message that it is only traced when the block is valid but not adopted).